modules/gateway/nodes.go

```diff
@@ -93,13 +93,15 @@ func (g *Gateway) relayNode(conn modules.PeerConn) error {
 	}
 	// add node
 	id := g.mu.Lock()
-	defer g.mu.Unlock(id)
 	if err := g.addNode(addr); err != nil {
+		g.mu.Unlock(id)
 		return err
 	}
 	g.save()
+	g.mu.Unlock(id)
```

Comment:
if you can't do that, then you should really wrap this in a func, which Luke is not a fan of but I very much dislike seeing locking happing across branches.

```go
err := func() error {
    id := g.mu.Lock()
    if err := g.addNode(addr); err != nil {
        return error
    }
   err = g.save()
   if err != nil {
        return err
    }
    return nil
}()
if err != nil {
    return err
}
```

But, really I think this can be avoided just be rethinking how the error handling should be happening here.